File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        vector<ListNode*> nodes;
        ListNode* curr = head;
        while (curr) {
            nodes.push_back(curr);
            curr = curr->next;
        }

        int n = nodes.size();

        ListNode* dummy = new ListNode(0);
        curr = dummy;
        int curr_start = 0;

        for (int i = k - 1; i < n; i += k) {
            curr_start = i;
            for (int j = 0; j < k; j++) {
                int idx = i - j;
                curr->next = nodes[idx];
                curr = curr->next;
            }
        }
        if (curr_start < n - 1) {
            curr->next = nodes[curr_start + 1];
        } else {
            curr->next = nullptr;
        }

        return dummy->next;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
### method1: AC but slow (mine)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        fast = head
        prev = None
        curr = head

        dummy = ListNode(next=head)
        last = dummy

        while True:
            cnt = k
            while cnt and fast:
                fast = fast.next
                cnt -= 1
            if cnt != 0:
                break

            cnt = k 
            while cnt:
                tmp = curr.next
                curr.next = prev
                prev = curr
                curr = tmp
                cnt -= 1
            
            tmp = last.next
            last.next = prev
            tmp.next = curr
            last = tmp

        return dummy.next


### method2: recursion

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = head
        cnt = k
        for i in range(k):
            if not curr: return head
            curr = curr.next

        curr = head
        prev = None
        for i in range(k):
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        head.next = self.reverseKGroup(curr, k)
        return prev